oasis-reference/src/main/kotlin/com/empowerops/volition/ref_oasis/OptimizerEndPoint.kt

```diff
@@ -0,0 +1,284 @@
+        state = State.Idle
+    }
+
+    override fun syncAll() {
+        GlobalScope.launch {
+            for (simName in modelService.simByName.keys) {
+                syncConfigFor(simName)
+            }
+        }
+    }
+
+    override fun updateNode(simName: String) {
+        GlobalScope.launch {
+            syncConfigFor(simName)
+        }
+    }
+
+    override fun cancelAll() {
+        GlobalScope.launch {
+            for ((name, sim) in modelService.simByName) {
+                val message = OASISQueryDTO.newBuilder().setCancelRequest(
+                        OASISQueryDTO.SimulationCancelRequest.newBuilder().setName(name)
+                ).build()
+
+                sim.input.onNext(message)
+            }
+        }
+    }
+
+    override fun cancel(name: String) {
+        GlobalScope.launch {
+            val message = OASISQueryDTO.newBuilder().setCancelRequest(
+                    OASISQueryDTO.SimulationCancelRequest.newBuilder().setName(name)
+            ).build()
+
+            modelService.simByName.getValue(name).input.onNext(message)
+        }
+    }
+
+    override fun disconnectAll() {
+       throw NotImplementedException()
+    }
+
+    override fun cancelAndStop() {
+        GlobalScope.launch {
+            stopOptimization()
+            for ((name, sim) in modelService.simByName) {
+                val message = OASISQueryDTO.newBuilder().setCancelRequest(
+                        OASISQueryDTO.SimulationCancelRequest.newBuilder().setName(name)
+                ).build()
+                sim.input.onNext(message)
+            }
+        }
+    }
+
+    private suspend fun syncConfigFor(simName: String) {
+        val sim = modelService.simByName.getValue(simName)
+        val message = OASISQueryDTO.newBuilder().setNodeStatusRequest(
+                OASISQueryDTO.NodeStatusUpdateRequest.newBuilder().setName(simName)
+        ).build()
+
+        sim.input.onNext(message)
+
+        val result = select<Either<Simulation, Message>> {
+            sim.update.onReceive { Either.Left(updateFromResponse(it)) }
+            sim.error.onReceive {
+                Either.Right(Message(it.name, "Error update simulation $simName due to ${it.message} :\n${it.exception}"))
+            }
+            onTimeout(Duration.ofSeconds(5).toMillis()) {
+                Either.Right(Message(simName, "Update simulation timeout. Please check simulation is registered and reponsive."))
+            }
+        }
+
+        if (result.isLeft()) {
+            modelService.updateSim(simName, result.left().get())
+        } else {
+            modelService.addMessage(result.right().get())
+        }
+
+    }
+
+    override fun stopOptimization(request: StopOptimizationCommandDTO, responseObserver: StreamObserver<StopOptimizationResponseDTO>) = responseObserver.consume {
+        stopOptimization()
+        StopOptimizationResponseDTO.newBuilder().setMessage("Stopped").setStopped(true).build()
```

Comment:
oof. So you do everything in parallel by default and just assume its successful?

why not do everything in sequential by default? It'l make it a lot easier to debug.

Reply:
Stop is not a suspend function in reference optimizer.